keps/sig-instrumentation/2831-kubelet-tracing/README.md

```diff
@@ -0,0 +1,373 @@
+
+```golang
+// KubeletConfiguration contains the configuration for the Kubelet
+type KubeletConfiguration struct {
+    metav1.TypeMeta `json:",inline"`
+    ------------
+
+    // +optional
+    // TracingConfiguration
+    TracingConfiguration componentbaseconfig.TracingConfiguration
+}
+
+```
+
+**NOTE** The TracingConfiguration will be defined in `kubernetes/component-base` repository
+```golang
+// TracingConfiguration specifies configuration for opentelemetry tracing
+type TracingConfiguration struct {
+    // +optional
+    // URL of the collector that's running on the node.
+    // Defaults to 0.0.0.0:4317
+    URL  *string `json:"url,omitempty" protobuf:"bytes,1,opt,name=url"`
+
+    // +optional
+    // SamplingRatePerMillion is the number of samples to collect per million spans.
+    // Defaults to 0.
+    SamplingRatePerMillion int32 `json:"samplingRatePerMillion,omitempty" protobuf:"varint,2,opt,name=samplingRatePerMillion"`
+}
+
+```
+
+## Design Details
+
+### Test Plan
+
+We will test tracing added by this feature with an integration test.  The
+integration test will verify that spans exported by the kubelet match what is
+expected from the request. We will also add an integration test that verifies
+spans propagated from kubelet to API server match what is expected from the request.
+
+### Graduation Requirements
+
+Alpha
+
+- [] Implement tracing of incoming and outgoing gRPC, HTTP requests in the kubelet
+- [] Integration testing of tracing
+
+Beta
+
+- [] Publish examples of how to use the OT Collector with kubernetes
+- [] Allow time for feedback
+- [] Revisit the format used to export spans.
+
+GA
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+  - [X] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: KubeletTracing
+    - Components depending on the feature gate: kubelet
+  - [X] Other
+    - Describe the mechanism: **Kubelet Configuration TracingConfiguration**
+      - To disable tracing entirely, do not enable the feature gate `KubeletTracing`.
+      - If KubeletTracing is enabled but no TracingConfiguration is provided, the sampling rate default value is 0.
+      - If KubeletTracing is enabled and sampling rate is 0:
+        - a trace context with sampled=true will still cause traces to be generated.
+        - kubelet will attempt to connect via otlp protocol to export traces.
+        - trace contexts will be propagated in passthrough mode. As stated in [OpenTelemetry documentation](https://github.com/open-telemetry/opentelemetry-go/tree/main/example/passthrough#passthrough-setup-for-opentelemetry), `the default TracerProvider implementation returns a 'Non-Recording' span that keeps the context of the caller but does not record spans`.
```

Comment:
"passthrough" mode means it won't generate any spans or connect via OTLP protocol.